modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/ContextualTransactionValidator.scala

```diff
@@ -0,0 +1,265 @@
+              case canOverride @ CanOverride(_) =>
+                (canOverride, txs.filterNot { case (ordinal, _) => ordinal > transaction.ordinal }.some).asRight
+              case CannotOverride(_, existing, current) => Conflict(transaction.ordinal, existing.hash, current.hash).asLeft
+            }
+          case None => (NoConflict(transaction), none).asRight
+        }
+
+      private def validateLastTransactionRef(
+        transaction: Hashed[Transaction],
+        txs: Option[SortedMap[TransactionOrdinal, StoredTransaction]],
+        lastProcessedTransactionRef: TransactionReference
+      ): Either[ContextualTransactionValidationError, Hashed[Transaction]] =
+        if (transaction.parent.ordinal < lastProcessedTransactionRef.ordinal)
+          ParentOrdinalLowerThenLastProcessedTxOrdinal(transaction.parent.ordinal, lastProcessedTransactionRef.ordinal).asLeft
+        else if (transaction.parent === lastProcessedTransactionRef) {
+          transaction.asRight
+        } else {
+          val hasValidParent = txs.exists(_.exists {
+            case (parentOrdinal, parentTx) => parentTx.ref === transaction.parent && parentOrdinal > lastProcessedTransactionRef.ordinal
+          })
+
+          if (hasValidParent) transaction.asRight else HasNoMatchingParent(transaction.parent.hash).asLeft
+        }
+
+      private def validateNotLocked(
+        transaction: Hashed[Transaction]
+      ): Either[ContextualTransactionValidationError, Hashed[Transaction]] =
+        Either
+          .cond(
+            !lockedAddresses.contains(transaction.source),
+            transaction,
+            LockedAddressError(transaction.source)
+          )
+
+      private def calculateLastAllowedTransactionWithinCap(
+        transactions: List[Hashed[Transaction]],
+        balance: Balance,
+        lastMajorityTransactionOrdinal: SnapshotOrdinal,
+        currentOrdinal: SnapshotOrdinal
+      ): Option[Hashed[Transaction]] = {
+        val ordinalDistanceForBaseBalance = Math
+          .floor(transactionLimitConfig.timeToWaitForBaseBalance / transactionLimitConfig.timeTriggerInterval)
+          .toLong
+
+        @tailrec
+        def loop(
+          remainingTransactions: List[Hashed[Transaction]],
+          balance: Balance,
+          lastTxRefOrdinal: SnapshotOrdinal,
+          totalConsumedCap: Double,
+          lastAcceptedWithinCap: Option[Hashed[Transaction]]
+        ): Option[Hashed[Transaction]] =
+          remainingTransactions match {
+            case head :: tail =>
+              val consumedCap =
+                if (head.fee.value >= transactionLimitConfig.minFeeWithoutLimit.value) 0
+                else {
+                  val distancePassed = currentOrdinal.value - lastTxRefOrdinal.value
+                  val transactionsAllowedWithinInterval =
+                    if (balance.value > 0) balance.value.toDouble / transactionLimitConfig.baseBalance.value.toDouble else 0.0
```

Comment:
`baseBalance` is a `NonNegLong`, shouldn't we for that reason check if it's greater than 0 as well before attempting division?

Reply:
I guess dividing by 0 doesn't throw - if that's the case then this works fine.